Initial documentation commit

fhashim opened the pull request:
@ta-oliver I've made an initial commit by creating a directory for documentation. Please refer to `infertrade/documentation/build/latex/infertrade.pdf` to view the pdf version of documentation. Also you can refer to `infertrade/documentation/build/html` to view documentation in HTML. 
Please note that I've started working with utilities module of the package. Your comments on progress in documentation would be helpful.

ta-oliver commented:
Many thanks, @fhashim .

Can you:

1. Add the required dependencies for auto-generation to requirements-dev.txt.
2. Add guidance on the steps needed to (re)generate the documentation. (e.g. "run `make help`").
3. Add the PDF (and perhaps html if not too messy) to version control so users can read the documentation without needing to run the makefile. GitIgnore any files that are not needed for final usage of the documentation.

fernando-az-infertrade commented on documentation/source/conf.py:
```diff
@@ -0,0 +1,56 @@
+author = 'Furqan Hashim'
+
+# The full version, including alpha/beta/rc tags
+release = '0.0.1'
```
maybe 
```
import _version
release = _version.__version__
```
?

fhashim replied:
@fernando-az-infertrade I've made suggested changes in commit [ef668a3](https://github.com/ta-oliver/infertrade/pull/103/commits/ef668a326713e89d3db2df709417dbcb48c654cd)

fernando-az-infertrade replied:
@fhashim , thank you.

fernando-az-infertrade commented:
I've never worked with sphinx, but this looks good so far :+1:

ta-oliver commented on documentation/source/conf.py:
```diff
@@ -0,0 +1,56 @@
+
+# -- Project information -----------------------------------------------------
+
+project = 'Infertrade'
+copyright = '2021, Furqan Hashim'
+author = 'Furqan Hashim'
```
```suggestion
project = 'InferTrade'
copyright = '2021, InferStat Ltd'
author = 'InferStat'
```

ta-oliver commented on documentation/source/index.rst:
```diff
@@ -0,0 +1,20 @@
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to Infertrade's documentation!
```
```suggestion
Welcome to the InferTrade package documentation
```

fhashim commented:
@ta-oliver I've completed the following pointers 

> Add the required dependencies for auto-generation to requirements-dev.txt.

I've added the Sphinx and pdflatex version required in requirements-dev.txt to build PDF & HTML of documentation.

>Add the PDF (and perhaps html if not too messy) to version control so users can read the documentation without needing to run the makefile. GitIgnore any files that are not needed for final usage of the documentation.

Documentation folder only contains the updated pdf documentation generated after taking into account the 5 commits that you've made.

ta-oliver commented:
@fhashim - many thanks.

I would suggest the following adjustments would be needed to get Sphinx to generate a good document:

1. We could move the Apache licence statements from a docstring to single line comments, as suggested [here]( https://stackoverflow.com/questions/17209712/how-do-i-get-sphinx-to-ignore-the-gpl-notice-in-my-docstring), so they don'[t appear everywhere in the auto-doc.

2. We should incorporate the existing readme.md into the auto-document.

3. Consider removing the excess blank pages?

fhashim commented:
@ta-oliver your suggestions were helpful. I've made the suggested changes. Please review and comment.
I would like to suggest that if we can modify the doc strings in our files (e.g. [infertrade/utilities/operations.py](https://github.com/fhashim/infertrade/blob/documentation/infertrade/utilities/operations.py)) as per google style as shown in code below. ([Reference](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html))
```
def func(arg1, arg2):
    """Summary line.

    Extended description of function.

    Args:
        arg1 (int): Description of arg1
        arg2 (str): Description of arg2

    Returns:
        bool: Description of return value

    """
    return True
```

fernando-az-infertrade commented:
@fhashim, I'm not sure if you've been added to SemaphoreCI yet, but it's currently (since commit 4abffd9) complaining about the following:

> ERROR: Cannot install -r requirements-dev.txt (line 4) and sphinx==4.0.2 because these package versions have conflicting dependencies.
> 
> The conflict is caused by:
>     The user requested sphinx==4.0.2
>     myst-parser 0.14.0 depends on sphinx<4 and >=2.1
> 
> To fix this you could try to:
> 1. loosen the range of package versions you've specified
> 2. remove package versions to allow pip attempt to solve the dependency conflict
> 
> ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
> WARNING: You are using pip version 20.3.3; however, version 21.1.2 is available.
> You should consider upgrading via the '/home/runner/.pyenv/3.7/bin/python3.7 -m pip install --upgrade pip' command.

fhashim commented:
@ta-oliver thanks for the assistance I've made the fix. [requirements-dev.txt ](https://github.com/fhashim/infertrade/blob/documentation/requirements-dev.txt) is now updated to fix requirements issue.

ta-oliver commented:
We can exclude this file?

![image](https://user-images.githubusercontent.com/29981664/120339062-b33b2400-c2ec-11eb-8812-1d51b65d2233.png)

ta-oliver commented:
@fernando-az-infertrade - would you be able to assist @fhashim and this PR by augmenting the auto-documentation with docsctrings etc?

fernando-az-infertrade commented:
> @fernando-az-infertrade - would you be able to assist @fhashim and this PR by augmenting the auto-documentation with docsctrings etc?

Yes, I'll look into making the docstrings and some other aspects more coherent as a whole.

ta-oliver commented:
I've moved all the Apache licences I could find to line comments, so the docstrings should be more relevant.

I think aligning format to Sphinx assumptions is good both for auto-documentation and will allow us to add the docs to https://readthedocs.org/

fernando-az-infertrade commented:
@fhashim , would you mind adding back to the branch the "make" files and .rst files removed in commit 5eb5d9d28ac42a3f3f2da8a8da288543aefbca4f?

I think this comment

> GitIgnore any files that are not needed for final usage of the documentation.

may have caused some confusion, but I believe the intent was to say that any intermediate files generated by the "make" tool should be added to .gitignore, and not that the "make" files themselves should be removed. Is that right, @ta-oliver ?

ta-oliver commented:
> @fhashim , would you mind adding back to the branch the "make" files and .rst files removed in commit [5eb5d9d](https://github.com/ta-oliver/infertrade/commit/5eb5d9d28ac42a3f3f2da8a8da288543aefbca4f)?
> 
> I think this comment
> 
> > GitIgnore any files that are not needed for final usage of the documentation.
> 
> may have caused some confusion, but I believe the intent was to say that any intermediate files generated by the "make" tool (if any) should be added to .gitignore, and not that the "make" files themselves should be removed. Is that right, @ta-oliver ?

Yes, exactly. We need to include all files that are necessary for building the documentation. We do not want to add to version control any files that are created during the documentation generation process.

fhashim commented:
@ta-oliver @fernando-az-infertrade I've added all the relevant files that are required to build the documentation.

fernando-az-infertrade commented:
@fhashim , Excellent! Thank you.

fhashim commented:
@ta-oliver @fernando-az-infertrade I've added a [README](https://github.com/fhashim/infertrade/blob/documentation/documentation/README.md) file which can guide on how to (re)generate documentation. I've kept it separate from the base [README](https://github.com/fhashim/infertrade/blob/documentation/README.md) as that is being used to build the documentation.

ta-oliver commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+
+To initialize the Sphinx source directory execute the following inside documentation:
+
+sphinx-quickstart
```
```suggestion
`sphinx-quickstart`
```

ta-oliver commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+
+Use sphinx-apidoc to generate reStructuredText files from source code by executing following command:
+
+sphinx-apidoc -f -o <path-to-output> <path-to-module>
```
```suggestion
`sphinx-apidoc -f -o <path-to-output> <path-to-module>`
```

ta-oliver commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+
+sphinx-apidoc -f -o <path-to-output> <path-to-module>
+
+Add the generated .rst file into index.rst toctree
```
Manually copy? What is required here?

fhashim replied:
@ta-oliver yes you need to manually copy the name of created .rst file to index.rst toc tree.

ta-oliver replied:
Could we just supply the desired location as the <path-to-output> argument?

ta-oliver commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+
+Finally to build document you can execute following command:
+
+make latexpdf
```
```suggestion
`make latexpdf`
```

ta-oliver commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+
+Additionally you can also build HTML by executing following command:
+
+make html
```
```suggestion
`make html`
```

fernando-az-infertrade commented on documentation/README.md:
```diff
@@ -0,0 +1,23 @@
+Ensure that requirements in requirements_dev are full filled.
+
+Make changes to conf.py as per the requirements.
+
+To initialize the Sphinx source directory execute the following inside documentation:
+
+sphinx-quickstart
+
+Currently the directory has been initialized. 
+
+Use sphinx-apidoc to generate reStructuredText files from source code by executing following command:
+
+sphinx-apidoc -f -o <path-to-output> <path-to-module>
+
+Add the generated .rst file into index.rst toctree
+
```
These steps are meant to guide the user through "re-creating" the conf.py and the .rst files that sphinx uses to build the final pdf/html files, correct?

@ta-oliver , was the idea of allowing users to re-build the documentation meant to also include these steps, or was it just meant to allow them to recreate the pdf/html based on the conf.py file and on the .rst files already provided in this repository? Because in the latter case, these steps can be cut out altogether.

ta-oliver replied:
Generally the users won't recreate the documentation, but just access a PDF, so the information should be primarily aimed at allow project contributors to be able to update the documentation automatically before each release.

fernando-az-infertrade commented on documentation/source/conf.py:
```diff
@@ -0,0 +1,65 @@
+#
+import os
+import sys
+from infertrade import _version
+
+sys.path.insert(0, os.path.abspath('../../infertrade/'))
```
```suggestion

sys.path.insert(0, os.path.abspath('../../infertrade/'))
import _version
```

Two things here:
(a) what I had thought of doing when I suggested inserting the `from infertrade import _version` line was simply to import the `_version.py` file that sits at the base of the infertrade package - which itself would, thanks to the `sys.path.insert` line, be placed at the beggining of the import path... however, this suggestion itself was incorrect, since the line would actually have to be just `import _version` (the `from infertrade` part would only be correct if the directory inserted in the path was the one *immediately above* the package);
(b) also, for the reasons previously indicated, the `sys.path.insert` line would have to sit above the the `import _version` line for this latter line to work (alternatively, we could just copy the `get_version` function from the `setup.py` file, as ta-oliver suggested in https://github.com/ta-oliver/infertrade/pull/103#issuecomment-852164432 ... @ta-oliver , do you have any preferences with regards to this?)

ta-oliver replied:
No strong preference - whichever would be easiest to understand and maintain?

fernando-az-infertrade replied:
@fhashim , just so you're aware, I'm committing this change.

fernando-az-infertrade commented on documentation/source/CONTRIBUTING.md:
```diff
@@ -0,0 +1,21 @@
+# Contributing to InferTrade
```
Suggestion, substitute the contents of both `documentation/source/CONTRIBUTING.md` and `documentation/source/README.md` for simple includes, as indicated here (in the first code block):  https://myst-parser.readthedocs.io/en/latest/using/howto.html#include-a-file-from-outside-the-docs-folder-like-readme-md

This way, we could avoid having to keep these documents synced up with the ones that exist at the root of the repository.

ta-oliver replied:
Yes, exactly - we shouldn't be duplicating the content but rather including by reference.

fernando-az-infertrade replied:
@fhashim , I've also commited commit 83fa5d91545154779cfbb12602fc212bc31c98a1 to address this suggestions.

fernando-az-infertrade commented:
@ta-oliver I've made tentative changes to the docstrings in a single file. Could you please check the last two commits two see if you have any notes?

ta-oliver commented on infertrade/utilities/operations.py:
```diff
@@ -136,10 +204,15 @@ def research_over_price_minus_one(x: Union[np.ndarray, pd.Series], shift: int) -
 
 class PricePredictionFromSignalRegression(TransformerMixin, BaseEstimator):
 
-    """Class for creating price predictions from signal values."""
+    """This class creates price predictions from signal values.
+
+    Args:
+        market_to_trade: The name of the column which contains the historical prices.
+
+    """
 
     def __init__(self, market_to_trade: str = None):
-        """We create by determining one input column as being the price to target."""
+        # We create by determining one input column as being the price to target.
```
Could keep as docstring for the construction method?

fernando-az-infertrade replied:
Do you mean just for this particular __init__ method or also for the two whose docstring just says "Trivial creation method"?

ta-oliver replied:
`The class constructor should be documented in the docstring for its __init__ method. Individual methods should be documented by their own docstring.`

Is the recommendation of [Pep257](https://www.python.org/dev/peps/pep-0257/)

ta-oliver replied:
So typically we can just say "Construction method for class X" in the docstring? Or similar?

fernando-az-infertrade replied:
:+1: I was going off a bad example then. I'll fix it.

ta-oliver commented:
> @ta-oliver I've made tentative changes to the docstrings in a single file. Could you please check the last two commits two see if you have any notes?

Generally looks fine. You could merge docstring changes to dev and then merge in dev changes to this branch, as docstring tidying can be immediately included (whereas we need PDF to be looking good and generation documentation to be sufficient to merge in this feature branch).

fernando-az-infertrade commented:
> > @ta-oliver I've made tentative changes to the docstrings in a single file. Could you please check the last two commits two see if you have any notes?
> 
> Generally looks fine. You could merge docstring changes to dev and then merge in dev changes to this branch, as docstring tidying can be immediately included (whereas we need PDF to be looking good and generation documentation to be sufficient to merge in this feature branch).

Will do.

ta-oliver commented:
>If we have PNG of logo it would be much easier to add logo to PDF.

![solo_image](https://user-images.githubusercontent.com/29981664/121071542-f8b59080-c7c7-11eb-8ada-82da81f291fc.png)
![InferTrade AppBar logo](https://user-images.githubusercontent.com/29981664/121071635-1b47a980-c7c8-11eb-92b8-7767054e31f0.png)

fhashim commented:
@ta-oliver is it fine if I replace the logo in README.md with the above logo?

ta-oliver commented:
> @ta-oliver is it fine if I replace the logo in README.md with the above logo?

Please do if it looks good.

fhashim commented:
@ta-oliver I've made changes and added logo to documentation by creating a separate .rst file for logo.  I also tried to use logo that is being used in README.md file. However, that logo was not looking good so I've used one of the 2 logos provided.

In order to use the logo, I had to create a folder and save images in those folder hence I've created an images directory in documentation folder. I've placed both the provided logos there. And I've referenced one of them in logo.rst file.

@ta-oliver @fernando-az-infertrade the logo appears on page 1 of the documentation. However, I believe we should have it on the introductory page the preamble part of documentation (i.e. the very first page of documentation). I am stuck at figuring out how can I add that in preamble. Any suggestions would be helpful. thanks

fernando-az-infertrade commented:
> I believe we should have it on the introductory page the preamble part of documentation (i.e. the very first page of documentation).

@fhashim , please check commit 1fcba39ed72ab6acb43053403bcfe8f63a0198f8 . Is this what you meant? Or did you mean something else?

ta-oliver commented:
Weird logo location is a minor problem that doesn't look great, but doesn't stop the documentation being usable.

Biggest problem is that the auto-documentation is not actually fully documenting the code in the repo. Only the utilities module is actually being documentation. So the main purpose of the documentation is not working. @fhashim - can you investigate the reasons for this and remediate? Once we have basic auto-documentation working we can try and improve the appearance of the doc.

ta-oliver commented:
I can relist main issues:

Many thanks, @fhashim - please see below my suggestions for getting the PDF documentation into a good appearance.

Required changes:

- [ ] Sections currently are only 'utilities'. Can we add all sections? (BIGGEST ISSUE)

![image](https://user-images.githubusercontent.com/29981664/120339920-71f74400-c2ed-11eb-99d6-c9202f22999c.png)

- [ ] Some of the docstrings could be improved so that the PDF makes more sense. E.g.:

![image](https://user-images.githubusercontent.com/29981664/120340199-b08cfe80-c2ed-11eb-966d-8747da48cb6e.png)

- [ ] Indices don't seem to be working. We could fix or remove?

![image](https://user-images.githubusercontent.com/29981664/120340264-c00c4780-c2ed-11eb-9ad4-4fe1947297c8.png)

- [ ] Python module index is incomplete. We may need to add docstrings and/or init files to fix autopopulation:

![image](https://user-images.githubusercontent.com/29981664/120340373-e16d3380-c2ed-11eb-831c-b2c007277391.png)

- [ ] Index format looks messy. Can we disable (people use find string) or improve via smaller fonts or similar?

![image](https://user-images.githubusercontent.com/29981664/120340521-03ff4c80-c2ee-11eb-830d-0d148b214924.png)

fhashim commented:
@ta-oliver we need to create .rst files for other directories to add them to documentation like what has been done for utilities.
As per my knowledge auto doc will only work once we have all the .rst for relevant modules in our documentation directory. I will start adding more modules into documentation.

fhashim commented:
@fernando-az-infertrade thanks a lot with the help. This is exactly what I meant.

ta-oliver commented:
Sounds good.

fernando-az-infertrade commented:
@fhashim , I understand that you were trying to start small with the `utilities` package, but I think we can now take a leap and try to run the autodoc on the whole package. Right now, we'd be more interested in getting the documentation for the whole package into a sort of "rough draft" (that we could then continuously improve), rather than getting everything to be perfect on the first go.

So I'd suggest you try to run Sphinx on the whole package and see if you can come up with something that looks presentable - it doesn't have to be perfect, as we'll work out the kinks later.

Also, if you run into any problems that should be eventually fixed but don't prevent other progress in the work you're doing, feel free to just edit your initial comment on this PR by adding a note saying "such and such needs to be fixed before merge", and then continue doing whatever work you can. If you do happen to hit a problem you can't get past, just post a comment on this PR to let us know, and then, while that isn't fixed, start reading up on the concept of "coverage" (as it relates to software development) and then on the "coverage" Python package, so you'll have something to do in the meantime (unfortunately, ta-oliver and I are kind of swamped at the moment doing work in other libraries, so we can't tend to every problem immediately).

fhashim commented:
@ta-oliver  @fernando-az-infertrade I've added 3 modules of package Infertrade, Algos, Data & Utilities to documentation. Please review the updated documentation. Currently, I've just added all relevant packages to build up the entire documentation. If any thing is missing please suggest. 
As discussed we can start document modification once all relevant packages are added to documentation.

ta-oliver commented:
@fhashim - many thanks. Starting to look good.

- [ ] Can you look at the structure of each section and see if we can reorder to remove blank subsections or add some relevant info to the docstrings?

![image](https://user-images.githubusercontent.com/29981664/122207693-4d2cd000-ce9a-11eb-848d-a8ab5150e95a.png)

- [ ] The Apache licences are still being included.
- [ ] Functions are running off the right side of the screen (can we reduce font size?).

![image](https://user-images.githubusercontent.com/29981664/122207858-82392280-ce9a-11eb-82dc-db42fde477f0.png)

- [ ] Odd end section:

![image](https://user-images.githubusercontent.com/29981664/122207937-9bda6a00-ce9a-11eb-9418-18d975f837d0.png)

fhashim commented:
@ta-oliver please review updated documentation pdf. I've removed empty sections form documentation by modifying the rst files. I've also removed section "Indices and Tables" which seemed odd. I was looking for some automated way to remove empty sections, but couldn't fine one. Therefore I've modified the rst files.

I am working on other suggested changes.

ta-oliver commented:
@fhashim - looks much better now. Thank you for cleaning the document up.

@SaurabhBansalInferstat - please could you review the PR, try generating docs etc and see if you have any comments or suggestions?

SaurabhBansalInferstat commented:
> @fhashim - looks much better now. Thank you for cleaning the document up.
> 
> @SaurabhBansalInferstat - please could you review the PR, try generating docs etc and see if you have any comments or suggestions?

yes I am reviewing it and is there 

> @fhashim - looks much better now. Thank you for cleaning the document up.
> 
> @SaurabhBansalInferstat - please could you review the PR, try generating docs etc and see if you have any comments or suggestions?



> @fhashim - looks much better now. Thank you for cleaning the document up.
> 
> @SaurabhBansalInferstat - please could you review the PR, try generating docs etc and see if you have any comments or suggestions?

I am reviewing it and so far it looks good. Although I  will try to generate the docs if I can and then get back to you if I have any suggestions.
